Fix issue in determineUrlType with URL polyfill.

sgomes opened the pull request:
The polyfill behaves incorrectly in some cases, parsing partial URLs instead of throwing an exception, as expected.

This patch takes the polyfill bugs into account to still arrive at a correct result.

The polyfill should only be getting used in IE11 and some very old browsers.

#### Changes proposed in this Pull Request

* Handle `URL` polyfill bugs in `lib/url`'s `determineUrlType`.

#### Testing instructions

There are no easy testing instructions, as the bug isn't present anywhere in live Calypso at the moment, to my knowledge. The changes are easy to grok, however: an absolute URL should have a protocol, and that protocol should not be the invalid `':'`.

The unit tests for `determineUrlType` can be used to ensure that the non-polyfilled implementation remains correct, even with the change.

sgomes committed:
Fix issue in determineUrlType with URL polyfill.

The polyfill behaves incorrectly in some cases, parsing partial URLs
instead of throwing an exception, as expected.

This patch takes the polyfill bugs into account to still arrive at a
correct result.

sgomes commented:
Any chance of a review? While I am not aware of any instances of this bug in production, it still makes me uncomfortable.

sirreal commented:
Is there a way of forcing the test suite to run against the polyfill version? Something like two test files that import the test suite, one with jsdom and the other importing the polyfill?

sgomes commented:
> Is there a way of forcing the test suite to run against the polyfill version? Something like two test files that import the test suite, one with jsdom and the other importing the polyfill?

It's probably doable, but I'm not sure how I'd prevent Jest from trying to run the shared file on its own. Any ideas?

sgomes committed:
Fix yet another polyfill issue.

sgomes committed:
Test both native and polyfilled URL.

Also removes a few tests that fail in the polyfilled version but can
probably be ignored.

sgomes commented:
@sirreal: Looks like I found a way of having a common file that gets loaded by two actual test files. I'm now testing `determineUrlType` with both the native `node` implementation and the polyfilled one.

sirreal commented on client/lib/url/test/common/url-type.skip.js:
```diff
@@ -53,8 +45,6 @@ describe( 'determineUrlType', () => {
 		// From https://url.spec.whatwg.org/#urls
-		expect( determineUrlType( 'https://ex ample.org/' ) ).toBe( 'INVALID' );
 		expect( determineUrlType( 'https://example.com:demo' ) ).toBe( 'INVALID' );
-		expect( determineUrlType( 'http://[www.example.com]/' ) ).toBe( 'INVALID' );
```
Were these bad test cases?

sgomes replied:
They failed in the polyfill, and I felt they don't add much to the tests; they'd be testing the polyfill/native implementation themselves more than `determineUrlType`.

sirreal commented on client/lib/url/url-type.ts:
```diff
@@ -52,8 +53,11 @@ export function determineUrlType( url: URLString | URL ): URL_TYPE {
 
 	try {
 		// If we can parse the URL without a base, it's an absolute URL.
+		// The polyfill works differently, however, so we need to take that into account.
```
Add a link to the issue in the polyfill repo?

sgomes replied:
Good idea! 👍

sirreal reviewed (approved):
Generally makes sense, nice to get the native and polyfilled cases tested ✨ 

Seems good to ship 👍

sgomes committed:
Add link to polyfill bug.
